shell/browser/api/electron_api_app.cc

```diff
@@ -1103,35 +1077,70 @@ void App::OnFirstInstanceAck(
 
 // This function handles the user calling
 // the callback parameter sent out by the second-instance event.
-static void AckCallbackWrapper(
-    const ProcessSingleton::NotificationAckCallback& ack_callback,
+void App::AckCallbackWrapper(
+    const ProcessSingleton::SendAckCallback& ack_callback,
     gin::Arguments* args) {
+  if (!send_ack_) {
+    return;
+  }
+
   blink::CloneableMessage ack_message;
   args->GetNext(&ack_message);
   if (!ack_message.encoded_message.empty()) {
-    ack_callback.Run(&ack_message.encoded_message);
+    ack_to_send_ = std::vector<const uint8_t>(ack_message.encoded_message.begin(), ack_message.encoded_message.end());
+    ack_timer_.Start(FROM_HERE, base::Seconds(0),
+                      base::BindOnce(ack_callback, &ack_to_send_));
   } else {
-    ack_callback.Run(nullptr);
+    ack_timer_.Start(FROM_HERE, base::Seconds(0),
+                      base::BindOnce(ack_callback, nullptr));
   }
 }
 
 void App::OnSecondInstance(
     const base::CommandLine& cmd,
     const base::FilePath& cwd,
     const std::vector<const uint8_t> additional_data,
-    const ProcessSingleton::NotificationAckCallback& ack_callback) {
+    const ProcessSingleton::SendAckCallback& ack_callback) {
   v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
   v8::Locker locker(isolate);
   v8::HandleScope handle_scope(isolate);
   v8::Local<v8::Value> data_value =
       DeserializeV8Value(isolate, std::move(additional_data));
-  auto cb = base::BindRepeating(&AckCallbackWrapper, ack_callback);
+  auto cb = base::BindRepeating(&App::AckCallbackWrapper, base::Unretained(this), ack_callback);
   bool prevent_default =
       Emit("second-instance", cmd.argv(), cwd, data_value, cb);
-  if (!prevent_default) {
+  if (!prevent_default && send_ack_) {
     // Call the callback ourselves, and send back nothing.
-    ack_callback.Run(nullptr);
+    ack_timer_.Start(FROM_HERE, base::Seconds(0),
+                      base::BindOnce(ack_callback, nullptr));
+  }
+}
+
+bool App::NotificationCallbackWrapper(
+    const base::RepeatingCallback<
+        void(const base::CommandLine& command_line,
+             const base::FilePath& current_directory,
+             const std::vector<const uint8_t> additional_data,
+             const ProcessSingleton::SendAckCallback& ack_callback)>&
+        callback,
+    const base::CommandLine& cmd,
+    const base::FilePath& cwd,
+    const std::vector<const uint8_t> additional_data,
+    const ProcessSingleton::SendAckCallback& ack_callback) {
+  send_ack_ = !Browser::Get()->is_shutting_down();
```

Comment:
Question: can I call `Browser::Get()->is_shutting_down()` at this point, or do I have to call it at the end before returning?
Note that the next line is `Browser::Get()->is_ready()`.
If `is_shutting_down` must be called at the end, we'll probably need a much larger fix to send back the ack properly.